File: LIB/atomtool.f90
MODULE atomtool
  IMPLICIT NONE
  CONTAINS
  
!INTEGER FUNCTION cnum(chem)
!INTEGER FUNCTION onum(chem)
!SUBROUTINE molweight(chem,weight)
!SUBROUTINE getatoms(chem,ica,ih,init,io,ir,isul,ifl,ibr,icl)

! ======================================================================
! Return the # of carbon atoms in a molecule                    
! ======================================================================
INTEGER FUNCTION cnum(chem)
  IMPLICIT NONE

  CHARACTER(LEN=*), INTENT(IN) :: chem   !  chemical formula
  INTEGER :: ic, i, nc

  cnum = 0 ; ic = 0 
  nc=INDEX(chem,' ')
  DO i=1,nc
    IF  (chem(i:i)=='C' .AND. chem(i:i+1)/='Cl') ic=ic+1
    IF  (chem(i:i)=='c') ic=ic+1
  ENDDO
  cnum = ic
  
END FUNCTION cnum

! ======================================================================
! Return the # of C-O-C (ether) in a molecule.
! (e.g. to estimate the # of of nodes in chem  
! ======================================================================
INTEGER FUNCTION onum(chem)
  IMPLICIT NONE

  CHARACTER(LEN=*),INTENT(IN) :: chem  ! chemical formula
  INTEGER :: io, i,nc

  onum = 0  ;  io   = 0
  nc=INDEX(chem,' ')-1
  DO i=1,nc
    IF (chem(i:i+1)=='-O') io=io+1
  ENDDO
  onum = io

END FUNCTION onum

! ======================================================================
! PURPOSE: compute the molecular weight of the species provided as input
! ======================================================================
SUBROUTINE molweight(chem,weight)
  IMPLICIT NONE

  CHARACTER(LEN=*),INTENT(IN) :: chem
  REAL,INTENT(OUT) :: weight

  INTEGER  ica, iha, ina, ioa, ira, isa, ifl, ibr, icl

  CALL getatoms(chem,ica,iha,ina,ioa,ira,isa,ifl,ibr,icl)
 
  weight = ica*12.01 + iha*1.01 + ina*14.01 + ioa*16 + isa*32.07 + &
           ifl*19 + ibr*79.9 + icl*35.45

END SUBROUTINE molweight

! ======================================================================
! Find the number of each possible atom in the species CHEM,
! e.g. C's, H's, N's, O's, S's, F's, Cl's, Br's, and .'s  
! ======================================================================
SUBROUTINE getatoms(chem,ica,ih,init,io,ir,isul,ifl,ibr,icl)
  USE keyparameter, ONLY:mxring
  USE rjtool
  IMPLICIT NONE
	 
  CHARACTER(LEN=*),INTENT(IN) ::  chem ! formula of the species 
  INTEGER,INTENT(OUT) :: ica  ! # of carbon atoms
  INTEGER,INTENT(OUT) :: ih   ! # of hydrogen atoms
  INTEGER,INTENT(OUT) :: init ! # of nitrogen atoms
  INTEGER,INTENT(OUT) :: io   ! # of oxygen atoms
  INTEGER,INTENT(OUT) :: ir   ! # of radical dot
  INTEGER,INTENT(OUT) :: isul ! # of sulfur atoms
  INTEGER,INTENT(OUT) :: ifl  ! # of fluorine atoms
  INTEGER,INTENT(OUT) :: ibr  ! # of bromine atoms
  INTEGER,INTENT(OUT) :: icl  ! # of chlorine atoms

  INTEGER         nring, i, n, nc
  CHARACTER(LEN=LEN(chem)) :: tchem  ! a copy of chem
  INTEGER         rjs(mxring,2)      ! ring-join group pairs
  LOGICAL         loring

! initialize
! ----------
  tchem=chem  ; nc=INDEX(tchem,' ')
  ifl = 0  ;  ibr = 0  ;  icl = 0 
  ica = 0  ;  ih  = 0  ;  init= 0  ;  io  = 0
  isul= 0  ;  ir  = 0
  n  = 1

  loring=.FALSE.
  IF (INDEX(tchem,'C1')/=0)  loring=.TRUE.
  IF (INDEX(tchem,'Cd1')/=0)  loring=.TRUE.
  IF (INDEX(tchem,'c1')/=0)  loring=.TRUE.
  IF (INDEX(tchem,'-O1')/=0) loring=.TRUE.
  IF (loring) THEN
     nring=2  ! set max # of rings (since # of rings is unknown here)
    CALL rjsrm(nring,tchem,rjs)
  ENDIF
  
! start counting
! --------------
  i=nc
  loopchar: DO
    i=i-1
    IF (i<1) EXIT loopchar
    IF (tchem(i:i)=='2') n = 2
    IF (tchem(i:i)=='3') n = 3
    IF (tchem(i:i)=='4') n = 4
    IF (i>1) THEN
      IF (tchem(i-1:i)=='Cl') THEN
        icl = icl + n
        n = 1  
        i = i-1
      ENDIF
    ENDIF
    IF (tchem(i:i)=='F') THEN      ; ifl  = ifl+n  ; n=1
    ELSE IF (tchem(i:i+1)=='Br') THEN ; ibr  = ibr+n  ; n=1
    ELSE IF (tchem(i:i)=='H') THEN ; ih   = ih+n   ; n=1
    ELSE IF (tchem(i:i)=='N') THEN ; init = init+n ; n=1
    ELSE IF (tchem(i:i)=='O') THEN ; io   = io+n   ; n=1
    ELSE IF (tchem(i:i)=='S') THEN ; isul = isul+1 ; n=1
    ELSE IF (tchem(i:i)=='.') THEN ; ir = ir+n     ; n=1
    ELSE IF ((tchem(i:i+1)/='Cl') .AND. &
             ((tchem(i:i)=='C').OR.(tchem(i:i)=='c')) ) THEN
       ica = ica+1 ; n  = 1
    ENDIF
  ENDDO loopchar

END SUBROUTINE getatoms

SUBROUTINE oxnum(chem,numoxC)
  USE keyparameter, ONLY:mxring,mxnode,mxlgr
  USE rjtool
  USE stdgrbond
  USE toolbox     , ONLY:countstring
  IMPLICIT NONE

  CHARACTER(LEN=*),INTENT(IN) :: chem   ! formula of the species
  REAL,INTENT(OUT)            :: numoxC ! Mean number of oxidation of carbon

  INTEGER                     :: noC(mxnode),nC,ngr
  INTEGER                     :: bond(mxnode,mxnode),dbflg,nring
  CHARACTER(LEN=mxlgr)        :: group(mxnode)
  INTEGER                     :: rjg(mxring,2) ! ring-join group pairs
  INTEGER                     :: i,j,nfunc

  numoxC=0. ; noC(:)=0 ; nC=0

  CALL grbond(chem,group,bond,dbflg,nring)
  ngr=COUNT(group/=' ')
  CALL rjgrm(nring,group,rjg)

  DO i=1,ngr
    IF ((group(i)(1:1)/='C').AND.(group(i)(1:1)/='c')) CYCLE
    nC=nC+1
    IF      (group(i)(1:3)=='CH3')  THEN ; noC(i)= -3
    ELSE IF (group(i)(1:3)=='CH2')  THEN ; noC(i)= -2
    ELSE IF (group(i)(1:2)=='CH')   THEN ; noC(i)= -1
    ELSE IF (group(i)(1:2)=='cH')   THEN ; noC(i)= -1
    ELSE IF (group(i)(1:3)=='CHO')  THEN ; noC(i)=  1
    ELSE IF (group(i)(1:2)=='CO')   THEN ; noC(i)=  2
    ENDIF

    nfunc=countstring(group(i),'(O') ; IF (nfunc/=0) noC(i)=noC(i)+1
    nfunc=countstring(group(i),'(N') ; IF (nfunc/=0) noC(i)=noC(i)+1

    DO j=1,ngr ; IF (bond(i,j)==3) noC(i)=noC(i)+1 ; ENDDO
  ENDDO

  numoxC=(SUM(noC(:)))/REAL(nC)

END SUBROUTINE oxnum

END MODULE atomtool

File: LIB/atomtool.c
/* Atom counting and manipulation tools for GECKO-A */
#include "atomtool.h"
#include "keyparameter.h"
#include "rjtool.h"
#include "toolbox.h"
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

/* Return the # of carbon atoms in a molecule */
int cnum(const char *chem) {
    int ic = 0;
    int nc = strcspn(chem, " ");
    if (nc == 0) nc = strlen(chem);
    
    for (int i = 0; i < nc; i++) {
        if (chem[i] == 'C' && !(i + 1 < nc && chem[i+1] == 'l')) {
            ic++;
        }
        if (chem[i] == 'c') {
            ic++;
        }
    }
    
    return ic;
}

/* Return the # of C-O-C (ether) in a molecule */
int onum(const char *chem) {
    int io = 0;
    int nc = strcspn(chem, " ");
    if (nc == 0) nc = strlen(chem);
    if (nc > 0) nc--;
    
    for (int i = 0; i < nc; i++) {
        if (i + 1 < (int)strlen(chem) && strncmp(chem + i, "-O", 2) == 0) {
            io++;
        }
    }
    
    return io;
}

/* Compute the molecular weight of the species */
void molweight(const char *chem, double *weight) {
    int ica, iha, ina, ioa, ira, isa, ifl, ibr, icl;
    
    getatoms(chem, &ica, &iha, &ina, &ioa, &ira, &isa, &ifl, &ibr, &icl);
    
    *weight = ica * 12.01 + iha * 1.01 + ina * 14.01 + ioa * 16.0 + 
              isa * 32.07 + ifl * 19.0 + ibr * 79.9 + icl * 35.45;
}

/* Find the number of each possible atom in the species CHEM */
void getatoms(const char *chem, int *ica, int *ih, int *init, int *io, 
              int *ir, int *isul, int *ifl, int *ibr, int *icl) {
    int nring, i, n, nc;
    char *tchem;
    int **rjs;
    bool loring;
    
    /* Initialize */
    nc = strcspn(chem, " ");
    if (nc == 0) nc = strlen(chem);
    
    tchem = (char *)malloc(strlen(chem) + 1);
    strcpy(tchem, chem);
    
    *ifl = 0; *ibr = 0; *icl = 0;
    *ica = 0; *ih = 0; *init = 0; *io = 0;
    *isul = 0; *ir = 0;
    n = 1;
    
    /* Check for rings */
    loring = false;
    if (strstr(tchem, "C1") != NULL) loring = true;
    if (strstr(tchem, "Cd1") != NULL) loring = true;
    if (strstr(tchem, "c1") != NULL) loring = true;
    if (strstr(tchem, "-O1") != NULL) loring = true;
    
    if (loring) {
        nring = 2;  /* set max # of rings */
        rjs = (int **)malloc(nring * sizeof(int *));
        for (i = 0; i < nring; i++) {
            rjs[i] = (int *)malloc(2 * sizeof(int));
        }
        rjsrm(nring, tchem, rjs, nring);
        
        for (i = 0; i < nring; i++) {
            free(rjs[i]);
        }
        free(rjs);
    }
    
    /* Start counting from the end of the string */
    i = nc;
    while (i > 0) {
        i--;
        
        if (tchem[i] == '2') n = 2;
        if (tchem[i] == '3') n = 3;
        if (tchem[i] == '4') n = 4;
        
        if (i > 0 && strncmp(tchem + i - 1, "Cl", 2) == 0) {
            *icl += n;
            n = 1;
            i--;
            continue;
        }
        
        if (tchem[i] == 'F') {
            *ifl += n;
            n = 1;
        } else if (i + 1 < nc && strncmp(tchem + i, "Br", 2) == 0) {
            *ibr += n;
            n = 1;
        } else if (tchem[i] == 'H') {
            *ih += n;
            n = 1;
        } else if (tchem[i] == 'N') {
            *init += n;
            n = 1;
        } else if (tchem[i] == 'O') {
            *io += n;
            n = 1;
        } else if (tchem[i] == 'S') {
            *isul += 1;
            n = 1;
        } else if (tchem[i] == '.') {
            *ir += n;
            n = 1;
        } else if ((tchem[i] == 'C' || tchem[i] == 'c') && 
                   !(i + 1 < nc && tchem[i+1] == 'l')) {
            *ica += 1;
            n = 1;
        }
    }
    
    free(tchem);
}

/* Calculate mean oxidation number of carbon - stub for now */
void oxnum(const char *chem, double *numoxC) {
    /* This function requires stdgrbond which hasn't been converted yet */
    /* For now, return 0 as a placeholder */
    (void)chem;  /* Unused */
    *numoxC = 0.0;
    
    /* TODO: Implement when stdgrbond is converted */
}
